### PR TITLE
Fix for repoGet request

### PR DESCRIPTION
**Pull request by maneesht**
Fix issue where the wrong cache was being updated by get. First noticed [here](https://github.com/firebase/firebase-ios-sdk/issues/8286)

### CODE REVIEW

File: packages/database/src/core/Repo.ts
Comment (IanWyszynski) [inline packages/database/src/core/Repo.ts:482]: Looking at this again, I suspect we may need to add a call to `repoRerunTranactions` call here. 

We do so in the data update handling code for a listener: https://github.com/firebase/firebase-js-sdk/blob/master/packages/database/src/core/Repo.ts#L400-L406.

Comment (IanWyszynski) [reply packages/database/src/core/Repo.ts:482]: (same thing for L504, after `applyTaggedQueryMerge`)

File: packages/database/src/core/Repo.ts
Comment (IanWyszynski) [reply packages/database/src/core/Repo.ts:497]: (another reason I'd prefer to move this synctree setup logic out of the result callback)

File: packages/database/test/exp/integration.test.ts
Comment (IanWyszynski) [inline packages/database/test/exp/integration.test.ts:113]: Perhaps it would be worth asserting that the snapshot is correct.

File: packages/database/src/core/Repo.ts
Comment (maneesht) [reply packages/database/src/core/Repo.ts:482]: I don't think we need to do this as `repoOnDataUpdate` does this for the listeners

File: packages/database/src/core/SyncTree.ts
Comment (IanWyszynski) [inline packages/database/src/core/SyncTree.ts:504]: Another comment here to explain the rational for splitting this out from syncTreeAddEventRegistration. Can be short.

File: packages/database/src/core/SyncTree.ts
Comment (IanWyszynski) [inline packages/database/src/core/SyncTree.ts:566]: It looks like you added this TODO? Could you either clarify this a bit more or get rid of ti?

File: packages/database/test/exp/integration.test.ts
Comment (IanWyszynski) [inline packages/database/test/exp/integration.test.ts:116]: I think these tests could be refactored into table-tests over a spec like this:

```
{
  "initialData": [{ name: 'child1' }, { name: 'child2' }],
  "initialLocation": <some path>,
  "listenerFunc": either onChild{Added, Changed, Removed} or onValue,
  "listenerQuery": <the query to listen to>,
  "expectedListenResults": [<array of snaps>] # fail if exactly these aren't received"
  "getQuery": <the query to get>,
   "expectedGetResults": <snap>, # fail if exactly this one isn't received.
  
}
```

Then we could easily add similar tests to the ones added here and check for various combinations of listens and gets (listen above get, below get, at get, etc.).

However, as I'm wrapping up this comment I suspect this refactor deserves it's own PR so feel free to ignore.

**Review by IanWyszynski — APPROVED** [review]
LGTM after some docs are added. Ignore the note about testing refactor (or take it as a suggestion for a followup PR).

**Comment by IanWyszynski** [comment]
We'll want to generate a changeset here: docs: https://github.com/firebase/firebase-js-sdk/pull/4408/files,  example: https://github.com/firebase/firebase-js-sdk/pull/4408/files.

File: .changeset/rotten-tables-brush.md
Comment (IanWyszynski) [inline .changeset/rotten-tables-brush.md:5]: nit: Fixed issue where `get()` updated cache incorrectly for non-default queries.

File: packages/database/src/core/Repo.ts
Comment (IanWyszynski) [inline packages/database/src/core/Repo.ts:480]: nit: s/listeners/listener

File: packages/database/src/core/Repo.ts
Comment (IanWyszynski) [inline packages/database/src/core/Repo.ts:494]: We should never get cancel events here, but maybe we should turn this into a warning log just in case.

```
if (cancels.length > 0) {
  repoLog(repo, 'ignoring unexpected cancel events in get for query ' + stringify(query));
}
```

File: packages/database/src/core/SyncTree.ts
Comment (IanWyszynski) [inline packages/database/src/core/SyncTree.ts:439]: The asymmetry of `syncTreeRegisterQuery` without `syncTreeUnregisterQuery` might raise some eyebrows. I would either add:

```
export function syncTreeUnregisterQuery(syncTree: SyncTree, query: QueryContext) {
  const cancels = syncTreeRemoveEventRegistration(
    repo.serverSyncTree_,
    query,
    null
   );
   if (cancels.length > 0) {
     repoLog(repo, 'ignoring unexpected cancel events in get for query ' + stringify(query));
   }
}
```

or note the reason we don't have `syncTreeUnregisterQuery` in the comments. I'm fine with either.

Comment (maneesht) [reply packages/database/src/core/SyncTree.ts:439]: I think this additional function makes sense

Comment (maneesht) [reply packages/database/src/core/SyncTree.ts:439]: Ended up just keeping it in `repoGetValue` since we aren't logging in `SyncTree` and it didn't seem like the right place to put the function.

File: packages/database/test/exp/integration.test.ts
Comment (IanWyszynski) [inline packages/database/test/exp/integration.test.ts:108]: Lets also add tests where the onValue and get are done on paths above and below each other.

Also, maybe a test that checks that nothing strange happens when a listener is removed

File: packages/database/src/core/SyncTree.ts
Comment (IanWyszynski) [inline packages/database/src/core/SyncTree.ts:440]: I'm not sure if this is incorrect, but it's worth pointing out that the `syncTreeAddEventRegistration` for a default query will do the logic of L441 to L452.

I'm not sure why the [assertion](https://github.com/firebase/firebase-js-sdk/blob/cdada6c68f9740d13dd6674bcb658e28e68253b6/packages/database/src/core/SyncTree.ts#L857-L858) in `syncTreeApplyTaggedQueryOverwrite` doesn't fail in tests for default queries, shouldn't it be set in `syncTreeRegisterSyncPoint`?

Comment (IanWyszynski) [reply packages/database/src/core/SyncTree.ts:440]: Will the tests still pass if you push this branch down to:
```
if (!query._queryParams.loadsAllData()) {
  return syncTreeTagForQuery_(syncTree, query);
} else {
  return null;
}
```

File: yarn.lock
Comment (IanWyszynski) [inline yarn.lock:17102]: Is this intentional? Seems like we shouldn't need to pull in an older version.

File: packages/database/test/exp/integration.test.ts
Comment (IanWyszynski) [inline packages/database/test/exp/integration.test.ts:122]: Could you turn this into a table test and apply this test case to a few more interesting queries?

```
const queries = [
  query(testRef, limitToFirst(1)),
  query(testRef, startAt("child2")),
  ... etc
]

queries foreach { q => runTest(q) }
```

Some inspiration from the query integration tests: https://github.com/firebase/firebase-js-sdk/blob/master/packages/database-compat/test/query.test.ts#L38.

Comment (maneesht) [reply packages/database/test/exp/integration.test.ts:122]: Will do.

File: yarn.lock
Comment (maneesht) [reply yarn.lock:17102]: Good catch. I assumed yarn would've resolved to the existing one installed in the repo

File: packages/database/src/core/SyncTree.ts
Comment (maneesht) [reply packages/database/src/core/SyncTree.ts:440]: Don't default queries not have tags?

Comment (IanWyszynski) [reply packages/database/src/core/SyncTree.ts:440]: Corrected.

File: yarn.lock
Comment (IanWyszynski) [reply yarn.lock:17102]: Looks like this is still here?

**Comment by IanWyszynski** [comment]
Latest commit w new tests LGTM. @jsdt could you have a look for code-owner approval?

**Review by IanWyszynski — APPROVED** [review]

**Review by jsdt — APPROVED** [review]

File: packages/database/package.json
Comment (hsubox76) [inline packages/database/package.json:60]: Can we avoid bringing in an extra dep here? I see in app-check we just implemented a function ourselves, is this good enough? https://github.com/firebase/firebase-js-sdk/blob/6f0049e66064809ae990a2d9461e28b2d6d08d19/packages/app-check/src/util.ts#L43

If it is, maybe we can extract it into `@firebase/util` and share it between app-check and database (maybe in a later PR, just have a duplicate implementation in this one?)

Comment (maneesht) [reply packages/database/package.json:60]: Yeah, let me just reimplement it here and add a TODO

**Review by hsubox76 — APPROVED** [review]

**Pull request merged**
